### PR TITLE
[9.0][PORT] account_analytic_required

### PR DESCRIPTION
**Pull request by yvaucher**

### CODE REVIEW

File: account_analytic_required/tests/test_account_analytic_required.py [hunk 1]
```diff
@@ -1,85 +1,86 @@
...
 from openerp.tests import common
-from openerp.osv import orm
+from openerp import exceptions
 
 
-class test_account_analytic_required(common.TransactionCase):
+class TestAccountAnalyticRequired(common.TransactionCase):
 
     def setUp(self):
         super(test_account_analytic_required, self).setUp()
```
Comment (cyrilgdn) [inline account_analytic_required/tests/test_account_analytic_required.py]: s/test_account_analytic_required/TestAccountAnalyticRequired/

So Travis is not happy.

File: account_analytic_required/models/account.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                              move_line.account_id.name,
+                              move_line.name,
+                              move_line.analytic_account_id.code,
+                              move_line.analytic_account_id.name)
+
+    @api.multi
+    def _check_analytic_required(self):
+        return not self._check_analytic_required_msg()
+
+    _constraints = [(_check_analytic_required,
```
Comment (guewen) [inline account_analytic_required/models/account.py]: is there a reason for not using `@api.constrains`?

**Comment by guewen** [comment]
Fixed the class name here: https://github.com/yvaucher/account-analytic/pull/1

File: account_analytic_required/models/account.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    @api.model
+    def _get_analytic_policy(self, account):
+        """ Extension point to obtain analytic policy for an account """
+        return account.user_type_id.analytic_policy
+
+    @api.multi
+    def _check_analytic_required_msg(self):
+        for move_line in self:
+            if move_line.debit == 0 and move_line.credit == 0:
```
Comment (alexis-via) [inline account_analytic_required/models/account.py]: Maybe we should use float_is_zero here...

**Comment by alexis-via** [comment]
same remarks as guewen about the use of api.constrains ; otherwise :+1:

**Comment by sbidoul** [comment]
@yvaucher do you plan to do the final touches to this one? (api.constrains + float_is_zero)

**Comment by yvaucher** [comment]
@sbidoul will work on it this week

**Comment by yvaucher** [comment]
fixed.

Travis is failing due to https://github.com/OCA/pylint-odoo/issues/48

**Comment by sbidoul** [comment]
Is it just me or there is no menu to manage account types?

**Comment by sbidoul** [comment]
Anyway, :+1: (code review and functional test on runbot).

Thanks for this work!

**Comment by yvaucher** [comment]
@sbidoul indeed there is no menuitem to access the view for account types and the field account type on the account is with widget type selection thus you have no way to enter an account type without creating the menuitem. Would it be better to restore the menuitem in this module or is it a requirement we can see in other modules and thus we need it in a separate module ?

**Comment by sbidoul** [comment]
@yvaucher we can leave that menu to another module if needed. I see a case for not letting end user fiddling with account types.

**Comment by alexis-via** [comment]
Can someone with the super-powers merge this ?

**Comment by rafaelbn** [comment]
:open_mouth: I would like to test it but I cannot find the menu por Account Types in v9 !

**Comment by sbidoul** [comment]
@rafaelbn there is no menu indeed. See https://github.com/OCA/account-analytic/pull/59#issuecomment-232082642 above.

To test, you can manually change the account.account view to remove widget="selection" on user_type, so you get a way to access the account.type form view.

**Comment by rafaelbn** [comment]
Sorry. Tested :+1:

File: account_analytic_required/README.rst [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/account-analytic/issues>`_. In case of trouble, please
+check there if your issue has already been reported. If you spotted it first,
+help us smashing it by providing a detailed and welcomed feedback.
+
+Credits
+=======
+
+Author
```
Comment (pedrobaeza) [inline account_analytic_required/README.rst]: You should remove this distinction. Everyone is a contributor. Alexis was the first contributor of a long list.

File: account_analytic_required/README.rst [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+------------
+* Stéphane Bidoul
+* Sefan Rijnhart
+* Laetitia Gangloff
+* Luc De Meyer, Noviat <info@noviat.com>
+* Yannick Vaucher <yannick.vaucher@camptocamp.com>
+
+Maintainer
+----------
+.. image:: http://odoo-community.org/logo.png
```
Comment (pedrobaeza) [inline account_analytic_required/README.rst]: https (in all places)

File: account_analytic_required/models/account.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+             "there is no analytic account ; if you select 'Never', the "
+             "accountant will get an error message if an analytic account "
+             "is present.")
+
+
+class AccountMoveLine(models.Model):
+    _inherit = "account.move.line"
+
+    @api.model
+    def _get_analytic_policy(self, account):
```
Comment (pedrobaeza) [inline account_analytic_required/models/account.py:38]: Use a related field instead this ugly trick

File: account_analytic_required/models/account.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+    @api.model
+    def _get_policies(self):
+        """This is the method to be inherited for adding policies"""
+        return [('optional', 'Optional'),
+                ('always', 'Always'),
+                ('never', 'Never')]
+
+    analytic_policy = fields.Selection(
+        lambda self, *args, **kwargs: self._get_policies(*args, **kwargs),
```
Comment (pedrobaeza) [inline account_analytic_required/models/account.py]: You should put directly the list of tuples here, as new API allows to extend this list. Even if you use a method, it should be declared this way (not with lambda):

```
analytic_policy = fields.Selection(_get_policies, ...)
```

**Comment by pedrobaeza** [comment]
We need a way to set the policy that this module adds:
- Add a menu here.
- Depends on other module that does it.

But we cannot merge this module as is.

**Comment by sbidoul** [comment]
@pedrobaeza to me the module is perfectly usable as is because in most projects we provide account types through a data module and users don't change them.

**Comment by pedrobaeza** [comment]
If you provide that data module, you need to depend on this one, or get a warning when it's not installed, so I don't see that as the correct option. One must be able to change this policy according business rules, not localization rules.

**Comment by sbidoul** [comment]
I'm not saying an account types menu is not useful, or that it must be set by localisation or business rule or...

I just say this PR is useful as is, because I have a project where I use it right know, where the analytic policies are set in a data module, and I don't need a menu to allow the user to manage the analytic policies.

So to me there is no reason to block this PR.

**Comment by pedrobaeza** [comment]
But it's for me, because this module has no value for me if we don't have that menu. That's why I'm asking for it.

**Comment by adrienpeiffer** [comment]
:+1: (Code review and functionnal test)

**Comment by rafaelbn** [comment]
Hi @yvaucher just a question, is it so complicated to add a menu to make it more useful for final users? Thanks

**Comment by yvaucher** [comment]
@rafaelbn I'll do it once I'm there in my todo list I just didn't add one as I have no idea if this was a specific requirement for this module or it should have been done in a base module. But I'll do it there and I will let other do the extract if needed then.

**Comment by rafaelbn** [comment]
Thanks for considering @yvaucher

File: account_analytic_required/models/account.py (hunk 6 above)
Comment (yvaucher) [inline account_analytic_required/models/account.py:38]: I think this is there for inheritance purpose that's what I understood from "Extension point" in the docstring.

File: account_analytic_required/README.rst (hunk 4 above)
Comment (yvaucher) [inline account_analytic_required/README.rst]: I kept original credits as I don't want to change them without @alexis-via approval.

File: account_analytic_required/models/account.py (hunk 6 above)
Comment (pedrobaeza) [inline account_analytic_required/models/account.py:38]: But now you don't need it with the new `selection_add` option

Comment (yvaucher) [inline account_analytic_required/models/account.py:38]: This is not a selection function you mix up with your other comment on lambda, right ?

**Comment by yvaucher** [comment]
PR has been fixed and is ready.

I answered the 2 open comments.

File: account_analytic_required/README.rst (hunk 4 above)
Comment (alexis-via) [inline account_analytic_required/README.rst]: @yvaucher you can change this, no problem !

**Comment by sbidoul** [comment]
:+1: 

@pedrobaeza @rafaelbn ok for you now?

**Comment by rafaelbn** [comment]
Great, many thanks! :+1:

**Comment by pedrobaeza** [comment]
:+1: Thanks

**Pull request merged**